fpm/src/fpm_sources.f90

```diff
@@ -0,0 +1,504 @@
+                             ['iso_c_binding  ', &
+                              'iso_fortran_env']
+
+type srcfile_ptr
+    ! For constructing arrays of src_file pointers
+    type(srcfile_t), pointer :: ptr => null()
+end type srcfile_ptr
+
+type srcfile_t
+    ! Type for encapsulating a source file 
+    !  and it's metadata
+    character(:), allocatable :: file_name
+        ! File path relative to cwd
+    character(:), allocatable :: exe_name
+        ! Name of executable for FPM_UNIT_PROGRAM
+    logical :: is_test = .false.
+        ! Is executable a test?
+    type(string_t), allocatable :: modules_provided(:)
+        ! Modules provided by this source file (lowerstring)
+    integer :: unit_type = FPM_UNIT_UNKNOWN
+        ! Type of program unit
+    type(string_t), allocatable :: modules_used(:)
+        ! Modules USEd by this source file (lowerstring)
+    type(string_t), allocatable :: include_dependencies(:)
+        ! Files INCLUDEd by this source file
+    type(srcfile_ptr), allocatable :: file_dependencies(:)
+        ! Resolved source file dependencies
+
+    logical :: built = .false.
+    logical :: touched = .false.
+end type srcfile_t
+
+
+contains
+
+subroutine add_sources_from_dir(sources,directory,with_executables)
+    ! Enumerate sources in a directory
+    !
+    type(srcfile_t), allocatable, intent(inout), target :: sources(:)
+    character(*), intent(in) :: directory
+    logical, intent(in), optional :: with_executables
+
+    integer :: i, j
+    logical, allocatable :: is_source(:), exclude_source(:)
+    type(string_t), allocatable :: file_names(:)
+    type(string_t), allocatable :: src_file_names(:)
+    type(srcfile_t), allocatable :: dir_sources(:)
+
+    ! Scan directory for sources
+    call list_files(directory, file_names)
+    file_names = [(string_t(directory//'/'//file_names(j)%s),j=1,size(file_names))]
+
+    is_source = [(str_ends_with(lower(file_names(i)%s), ".f90") .or. &
+                  str_ends_with(lower(file_names(i)%s), ".c") .or. &
+                  str_ends_with(lower(file_names(i)%s), ".h"),i=1,size(file_names))]
+    src_file_names = pack(file_names,is_source)
+
+    allocate(dir_sources(size(src_file_names)))
+    allocate(exclude_source(size(src_file_names)))
+
+    do i = 1, size(src_file_names)
+
+        if (str_ends_with(lower(src_file_names(i)%s), ".f90")) then
+            dir_sources(i) = parse_f_source(src_file_names(i)%s)
+        end if
+
+        if (str_ends_with(lower(src_file_names(i)%s), ".c") .or. &
+            str_ends_with(lower(src_file_names(i)%s), ".h")) then
+            dir_sources(i) = parse_c_source(src_file_names(i)%s)
+        end if
+
+        ! Exclude executables unless specified otherwise
+        exclude_source(i) = (dir_sources(i)%unit_type == FPM_UNIT_PROGRAM)
+        if (dir_sources(i)%unit_type == FPM_UNIT_PROGRAM .and. &
+             present(with_executables)) then
+            if (with_executables) then
```

Comment:
I had to read this part twice to understand it. Maybe adding an `&` in the beginning of 97 would make this clearer.

Reply:
```suggestion
        if (dir_sources(i)%unit_type == FPM_UNIT_PROGRAM .and. &
                                        present(with_executables)) then
            if (with_executables) then
```
I normally use a bit more indentation in these situations (see above) to avoid the issue.
Is this any clearer? If not I'll put the extra ampersand on 97.

Reply:
I'm not a friend of hanging indentation in Fortran at all, but this is just my personal taste. Also I usually prefer to put the optional ampersand on the beginning of every continuation line, to enhance readability, again personal taste.